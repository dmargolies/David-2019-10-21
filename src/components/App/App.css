.App {
  margin: 0 auto;
  max-width: 960px;
  text-align: center;
}

.header {
  align-items: baseline;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  text-align: left;
}

.search {
  border: 1px solid black;
  height: 2em;
  flex-basis: 390px;
}

@media all and (max-width: 375px) {
  .header {
    flex-direction: column;
    align-items: stretch;
  }

  .search {
    flex-basis: 0;
  }
}
