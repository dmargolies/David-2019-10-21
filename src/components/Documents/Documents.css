.documents {
  display: flex;
  flex-wrap: wrap;
  margin: -12px;
}

@media all and (max-width: 631px) {
  .documents {
    flex-direction: column;
    align-items: stretch;
  }
}

.document {
  border: 1px solid black;
  box-sizing: border-box;
  flex-basis: 304px;
  margin: 12px;
  padding: 15px;
}

@media all and (max-width: 631px) {
  .document {
    flex-basis: 0;
  }
}

.info {
  display: flex;
  justify-content: space-between;
}

.name {
  text-align: left;
  margin-bottom: 8px;
}
