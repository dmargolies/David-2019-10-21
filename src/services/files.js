const BASE = '/api/files';

async function betterFetch(url, params) {
  const response = await fetch(url, params)
  const json = await response.json();

  if (!response.ok) {
    return Promise.reject({ ...json, status: response.status, statusText: response.statusText });
  }

  return json;
}

export function deleteFile(filekey) {
  return betterFetch(`${BASE}/${filekey}`, { method: 'DELETE' });
}

export function fetchFiles() {
  return betterFetch(BASE);
}

export function postFile(file) {
  const body = new FormData();
  body.append('file', file);

  return betterFetch(BASE,{ method: 'POST', body });
}

export function searchFiles(searchTerm) {
  return betterFetch(`${BASE}?q=${searchTerm}`);
}
