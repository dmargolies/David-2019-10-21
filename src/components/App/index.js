import App from './App';
export default App; // TODO enable es6 feature for this?
