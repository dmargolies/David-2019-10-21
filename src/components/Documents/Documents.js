import React from 'react';
import PropTypes from 'prop-types';
import './Documents.css';

import formatSize from 'utils/formatSize';

function Documents({ documents, onDelete }) {
  return (
    <section className="documents">
      {documents.map(({ filekey, filename, size }) => (
        <Document key={filekey} name={filename} size={size} onDelete={() => onDelete(filekey)} />
      ))}
    </section>
  );
};

Documents.propTypes = {
  documents: PropTypes.arrayOf(PropTypes.shape({
    filekey: PropTypes.string,
    filename: PropTypes.string,
    size: PropTypes.number
  })),
  onDelete: PropTypes.func
};

Documents.defaultProps = {
  documents: [],
  onDelete: () => {}
};

// TODO if we put any links, we need to sanitize, or they could make the filename "javascript: alert('abc')"
function Document({ name, size, onDelete }) {
  return (
    <div className="document">
      <h3 className="name">{name}</h3>
      <section className="info">
        <span>{formatSize(size)}</span>
        <button onClick={onDelete}>delete</button>
      </section>
    </div>
  );
}

export default Documents;
