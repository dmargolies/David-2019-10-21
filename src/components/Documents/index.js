import Documents from './Documents';
export default Documents; // TODO enable es6 feature for this?
