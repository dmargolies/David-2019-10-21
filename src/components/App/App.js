import React, { useState, useEffect } from 'react';
import './App.css';

import Documents from 'components/Documents';
import FileInput from 'components/FileInput';

import { deleteFile, fetchFiles, postFile, searchFiles } from 'services/files';
import formatSize, { ONE_MEGABYTE } from 'utils/formatSize';
import generateUploadError from 'utils/generateUploadError';

const TEN_MEGABYTES = ONE_MEGABYTE * 10;

const getErrorMessage = ({ msg, statusText }) => msg || statusText;

function App() {
  const [documents, setDocuments] = useState([]);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchFiles()
      .then(({ files, error }) => {
        setDocuments(files);
      })
      .catch(err => setError(getErrorMessage(err)));
  }, []);

  useEffect(() => {
    if (error) {
      window.alert(error);
      setError(null);
    }
  }, [error]);

  function handleUpload({ error, file }) {
    if (error) {
      return setError(generateUploadError(error, file));
    }

    postFile(file)
      .then(newDocument => setDocuments([...documents, newDocument]))
      .catch(err => setError(getErrorMessage(err)));
  }

  function handleDelete(filekey) {
    deleteFile(filekey)
      .then(() => setDocuments(documents.filter(file => file.filekey !== filekey)))
      .catch(err => setError(getErrorMessage(err)));
  }

  function handleSearch({ target: { value: searchTerm }}) {
    // TODO debounce this
    searchFiles(searchTerm)
      .then(({ files }) => setDocuments(files))
      .catch(err => setError(getErrorMessage(err)));
  }

  const totalSize = documents.reduce((acc, { size }) => acc + size, 0);

  return (
    <div className="App">
      <header className="header">
        <input
          className="search"
          type="text"
          placeholder="Search documents..."
          onChange={handleSearch}/>
        <FileInput
          accept="image/jpeg, image/png"
          maxSize={TEN_MEGABYTES}
          onUpload={handleUpload} />
      </header>

      <header className="header">
        <h1>{documents.length} documents</h1>
        <span>Total size: {formatSize(totalSize)}</span>
      </header>

      <Documents documents={documents} onDelete={handleDelete}/>
    </div>
  );
}

export default App;
