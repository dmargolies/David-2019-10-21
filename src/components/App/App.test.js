import React from 'react';
import { create, act } from 'react-test-renderer';
import { render, unmountComponentAtNode } from 'react-dom';
import { act as domAct, Simulate } from 'react-dom/test-utils';

import App from './App';
import Documents from 'components/Documents';
import { deleteFile, fetchFiles, searchFiles } from 'services/files';

jest.mock('services/files');

beforeEach(() => {
  fetchFiles.mockResolvedValue({
    files: [{ filename: 'f1', size: 300432 }, { filename: 'f2', size: 300212 }]
  });

  searchFiles.mockResolvedValue({
    files: [{ filename: 'f1', size: 300432 }]
  });

  deleteFile.mockResolvedValue();
});

describe('DOM-less tests ', () => {
  let testRenderer;
  beforeEach( async () => {
    await act(async () => {
      testRenderer = create(<App />);
    });
  });

  it('should render correctly', async () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
    expect(fetchFiles).toHaveBeenCalled();
  });

  it('should render correctly when there are no documents', async () => {
    fetchFiles.mockResolvedValue({ files: [] });

    expect(testRenderer.toJSON()).toMatchSnapshot();
    expect(fetchFiles).toHaveBeenCalled();
  });

  it('should remove a file once deleted', async () => {
    const testInstance = testRenderer.root;
    const firstDoc = testInstance.findByProps({ name: 'f1' });

    await act(async () => {
      firstDoc.props.onDelete();
    });

    expect(testInstance.findByType(Documents).props.documents)
      .toEqual([{ filename: 'f2', size: 300212 }]);
  });
});

describe('DOM tests', () => {
  let container;
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('should search for documents when the search input updates', async () => {
    await domAct(async () => {
      render(<App />, container);
    });

    const input = container.querySelector('input[type="text"]');

    await domAct(async () => {
      Simulate.change(input, {target: { value: 'abc' }});
    });

    expect(searchFiles.mock.calls[0][0]).toBe('abc');
  });
});
