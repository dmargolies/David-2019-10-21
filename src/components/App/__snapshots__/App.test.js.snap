// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM-less tests  should render correctly 1`] = `
<div
  className="App"
>
  <header
    className="header"
  >
    <input
      onChange={[Function]}
      placeholder="Search documents..."
      type="text"
    />
    <input
      accept="image/jpeg, image/png"
      hidden={true}
      onChange={[Function]}
      type="file"
    />
    <button
      onClick={[Function]}
    >
      UPLOAD
    </button>
  </header>
  <header
    className="header"
  >
    <h1>
      2
       documents
    </h1>
    <span>
      Total size: 
      601kB
    </span>
  </header>
  <section
    className="documents"
  >
    <div
      className="document"
    >
      <h3
        className="name"
      >
        f1
      </h3>
      <section
        className="info"
      >
        <span>
          300kB
        </span>
        <button
          onClick={[Function]}
        >
          delete
        </button>
      </section>
    </div>
    <div
      className="document"
    >
      <h3
        className="name"
      >
        f2
      </h3>
      <section
        className="info"
      >
        <span>
          300kB
        </span>
        <button
          onClick={[Function]}
        >
          delete
        </button>
      </section>
    </div>
  </section>
</div>
`;

exports[`DOM-less tests  should render correctly when there are no documents 1`] = `
<div
  className="App"
>
  <header
    className="header"
  >
    <input
      onChange={[Function]}
      placeholder="Search documents..."
      type="text"
    />
    <input
      accept="image/jpeg, image/png"
      hidden={true}
      onChange={[Function]}
      type="file"
    />
    <button
      onClick={[Function]}
    >
      UPLOAD
    </button>
  </header>
  <header
    className="header"
  >
    <h1>
      2
       documents
    </h1>
    <span>
      Total size: 
      601kB
    </span>
  </header>
  <section
    className="documents"
  >
    <div
      className="document"
    >
      <h3
        className="name"
      >
        f1
      </h3>
      <section
        className="info"
      >
        <span>
          300kB
        </span>
        <button
          onClick={[Function]}
        >
          delete
        </button>
      </section>
    </div>
    <div
      className="document"
    >
      <h3
        className="name"
      >
        f2
      </h3>
      <section
        className="info"
      >
        <span>
          300kB
        </span>
        <button
          onClick={[Function]}
        >
          delete
        </button>
      </section>
    </div>
  </section>
</div>
`;
